Assets/Scripts/Visuals/UiHierarchyVisuals.cs

```diff
@@ -71,17 +78,53 @@ private void MakeHierarchyItemsRecursive(UiBuilder uiBuilder, int level, IEnumer
 
                 var isExpanded = _hierarchyChangeSystem.IsExpanded(entity);
 
-                uiBuilder.HierarchyItem(entity.ShownName, level, entity.Children.Any(), isExpanded, style, new HierarchyItemHandler.UiHierarchyItemActions
+                uiBuilder.HierarchyItem(entity.ShownName, level, entity.Children.Any(), isExpanded, style,
+                    new HierarchyItemHandler.UiHierarchyItemActions
                 {
                     OnArrowClick = () => { _hierarchyChangeSystem.ClickedOnEntityExpandArrow(entity); },
                     OnNameClick = () => { _hierarchyChangeSystem.ClickedOnEntityInHierarchy(entity); }
-                });
+                }, isPrimarySelection);
 
                 if (isExpanded)
                 {
                     MakeHierarchyItemsRecursive(uiBuilder, level + 1, entity.Children);
                 }
             }
         }
+        
+        private void ExpandSelectedItem()
+        {   
+            var selectedEntity = _sceneDirectoryState.CurrentScene?.SelectionState.PrimarySelectedEntity;
+            _hierarchyChangeSystem.ExpandParents(selectedEntity);
+        }
+
+        private HierarchyItemHandler HierarchyPanelSelectedItem()
+        {
+            HierarchyItemHandler uiItemHandler = null;
+            foreach (var uiItem in scrollRect.content.GetChildren())
+            {
+                if (!uiItem.TryGetComponent<HierarchyItemHandler>(out var hierarchyItemHandler)) continue;
+
+                if (hierarchyItemHandler.primarySelection)
+                    uiItemHandler = hierarchyItemHandler;
+            }
+
+            return uiItemHandler;
+        }
+
+        private void ScrollPanelToSelectedItem()
```

Comment:
Can you change this, so the view only scrolls to the position, when the selected item would be outside the visible area?
I think, it might be confusing, when the hierarchy moves. So try to minimize it and only move it, when it is necessary to see the selected entity.